processor/src/tests/u32_ops.rs

```diff
@@ -0,0 +1,1009 @@
+use proptest::prelude::*;
+use rand_utils::rand_value;
+use vm_core::{Felt, StarkField};
+
+// CONSTANTS
+// ================================================================================================
+const U32_BOUND: u64 = 1_u64 << 32;
+const WORD_LEN: usize = 4;
+
+// HELPER FUNCTIONS
+// ================================================================================================
+
+/// This helper function tests that when the given assembly instruction is executed on the
+/// the provided inputs, it results in the provided outputs.
+fn test_execution(asm_op: &str, inputs: &[u64], outputs: &[u64]) {
+    let script = compile(format!("begin {} end", asm_op).as_str());
+
+    let inputs = build_inputs(inputs);
+    let trace = execute(&script, &inputs).unwrap();
+    let last_state = trace.last_stack_state();
+
+    let expected_state = build_stack_state(outputs);
+    assert_eq!(expected_state, last_state);
+}
+
+/// This helper function tests failures where the execution of a given assembly operation over the
+/// provided inputs is expected to panic. This function catches the panic and tests it against a
+/// provided string to make sure it contains the expected error string.
+fn test_execution_failure(asm_op: &str, inputs: &[u64], err_substr: &str) {
+    let script = compile(format!("begin {} end", asm_op).as_str());
+
+    let inputs = build_inputs(inputs);
+    assert_eq!(
+        std::panic::catch_unwind(|| execute(&script, &inputs).unwrap())
+            .err()
+            .and_then(|a| { a.downcast_ref::<String>().map(|s| s.contains(err_substr)) }),
+        Some(true)
+    );
+}
+
+/// This helper function tests failures where the compilation of a given assembly operation is
+/// expected to panic. This function catches the panic and tests it against a provided string to
+/// make sure it contains the expected error string.
+fn test_compilation_failure(asm_op: &str, err_substr: &str) {
+    assert_eq!(
+        std::panic::catch_unwind(|| compile(format!("begin {} end", asm_op).as_str()))
+            .err()
+            .and_then(|a| { a.downcast_ref::<String>().map(|s| s.contains(err_substr)) }),
+        Some(true)
+    );
+}
+
+/// This helper function tests a provided u32 assembly operation, which takes a single input, to
+/// ensure that it fails when the input is >= 2^32.
+fn test_input_out_of_bounds(asm_op: &str) {
+    test_execution_failure(asm_op, &[U32_BOUND], "FailedAssertion");
+}
+
+/// This helper function tests a provided u32 bitwise assembly operation which takes a parameter
+/// to ensure that it fails when that parameter is out of bounds above the maximum.
+fn test_param_out_of_bounds(asm_op_base: &str, over: u64) {
+    let build_asm_op = |param: u64| format!("{}.{}", asm_op_base, param);
+
+    // --- test b > max ---------------------------------------------------------------------------
+    test_compilation_failure(build_asm_op(over).as_str(), "parameter value");
```

Comment:
I wonder if most (all?) of these helper functions should be in the module one level higher. I'd imagine they'd be useful for other tests as well.

Reply:
I moved all but one up a level & updated the ```execute``` import for the other test files. The other tests could probably be updated to use the new helpers (especially ```test_execution```), but I left that alone for now.

Reply:
Thank you! Agreed - propagating `test_execution` should be done in another PR.